### PR TITLE
Add duckdb_enum_values C API function to get enum dictionary strings

### PR DESCRIPTION
**Pull request by karlseguin**
This adds a `duckdb_enum_values` function to the C-API. It populates the passed-in `char**` with the enum string values, preserving the insertion order. This allows a driver to accept a string value and map it to the internal integer representation. I plan on using in my Zig driver to allow string values to be appended to enum columns (using the existing `duckdb_append_data_chunk`).

### CODE REVIEW

File: src/include/duckdb.h [hunk 1]
```diff
@@ -1810,6 +1810,21 @@ The result must be freed with `duckdb_free`.
...
+/*!
+Retrieves the dictionary values from the enum.
+
+While the values array is owned by the caller, the individual string must be freed
+with `duckdb_free`.
+
+* type: The logical type object
+* values: The array to store the values in
+* length: The length of the values parameter
+* returns: The number of values written to values. This will be the min betwen
```
Comment (Tishj) [inline src/include/duckdb.h:1822]: between*

File: src/include/duckdb.h [hunk 2]
```diff
@@ -1810,6 +1810,21 @@ The result must be freed with `duckdb_free`.
...
+* the lenght parameter and the size of the enum dictionary size (i.e.
```
Comment (Tishj) [inline src/include/duckdb.h:1823]: length*

File: src/main/capi/logical_types-c.cpp [hunk 3]
```diff
@@ -204,6 +204,24 @@ char *duckdb_enum_dictionary_value(duckdb_logical_type type, idx_t index) {
 	return strdup(duckdb::StringValue::Get(value).c_str());
 }
 
+idx_t duckdb_enum_values(duckdb_logical_type type, char **values, idx_t length) {
```
Comment (Tishj) [inline src/main/capi/logical_types-c.cpp:207]: Can we add a check and a test for passing in `values` as NULL?

Perhaps the returned length should be an out-param instead and this function should return a `duckdb_state`?

File: src/include/duckdb.h (hunk 1 above)
Comment (karlseguin) [reply src/include/duckdb.h]: sorry, fixed.

File: src/main/capi/logical_types-c.cpp (hunk 3 above)
Comment (karlseguin) [reply src/main/capi/logical_types-c.cpp]: Ok, I changed this. Added a null check and test for it.

**Comment by Tishj** [comment]
```
D:\a\duckdb\duckdb\src\main\capi\logical_types-c.cpp(216,45): error C2065: 'string_t': undeclared identifier [D:\a\duckdb\duckdb\src\main\capi\duckdb_main_capi.vcxproj]
D:\a\duckdb\duckdb\src\main\capi\logical_types-c.cpp(216,37): error C2672: 'duckdb::FlatVector::GetData': no matching overloaded function found [D:\a\duckdb\duckdb\src\main\capi\duckdb_main_capi.vcxproj]
  D:\a\duckdb\duckdb\src\include\duckdb\common\types\vector.hpp(329,19):
  could be 'T *duckdb::FlatVector::GetData(duckdb::Vector &)'
  	D:\a\duckdb\duckdb\src\main\capi\logical_types-c.cpp(216,45):
  	'duckdb::FlatVector::GetData': invalid template argument for 'T', type expected
  D:\a\duckdb\duckdb\src\include\duckdb\common\types\vector.hpp(325,25):
  or       'const T *duckdb::FlatVector::GetData(const duckdb::Vector &)'
  	D:\a\duckdb\duckdb\src\main\capi\logical_types-c.cpp(216,45):
  	'duckdb::FlatVector::GetData': invalid template argument for 'T', type expected
  
D:\a\duckdb\duckdb\src\main\capi\logical_types-c.cpp(221,16): error C3536: 'strings': cannot be used before it is initialized [D:\a\duckdb\duckdb\src\main\capi\duckdb_main_capi.vcxproj]
D:\a\duckdb\duckdb\src\main\capi\logical_types-c.cpp(221,25): error C2109: subscript requires array or pointer type [D:\a\duckdb\duckdb\src\main\capi\duckdb_main_capi.vcxproj]
D:\a\duckdb\duckdb\src\main\capi\logical_types-c.cpp(222,47): error C3536: 'value': cannot be used before it is initialized [D:\a\duckdb\duckdb\src\main\capi\duckdb_main_capi.vcxproj]
```

**Comment by karlseguin** [comment]
I don't understand the build. I was able to run the tests and build it. But I changed `string_t` to `duckdb::string_t` which `duckdb_create_enum_type` in the hope that this is what was needed.

**Comment by Mytherin** [comment]
Thanks for the PR! Can't the same behavior be accomplished by using `duckdb_enum_dictionary_size` and `duckdb_enum_dictionary_value`?

In general we want to keep the C API lean and not add functions unnecessarily. It is not meant to be nice to use stand-alone but rather meant as a building block for other APIs, so if the functionality is available elsewhere adding extra functions is not desirable.

**Comment by Giorgi** [comment]
@Mytherin What about adding a function that takes an enum member name and returns its numeric value? Similar to `duckdb_enum_dictionary_value` but reversed. Sure, it can be done with `duckdb_enum_dictionary_size` and `duckdb_enum_dictionary_value` but I think it's useful to have a dedicated function instead of calling `duckdb_enum_dictionary_value` in a loop.

**Comment by karlseguin** [comment]
You know...I didn't want to couple the size of the dictionary to the underlying enum value. I figured the two might not always be tightly coupled: what if  you were allowed to drop an enum value, then you'd have a hole. But I realize that might implementation has the same baked-in coupling by depending on the index of the array...So sorry.

**Comment by Mytherin** [comment]
No need to apologize.

The logical type contains the type after catalog resolution, so a drop/create of the enum value would not affect it as we are not going back to the catalog between these steps. As such that scenario will work correctly in both cases.

**Pull request closed without merging**